docs/proposals/apb-state-support.md

```diff
@@ -0,0 +1,59 @@
+# Abstract
+
+This proposal will outline a simple and secure method for allowing Service Bundle developers to set state during a Service Bundle
+action and then be able to access that state during a subsequent Service Bundle action.
+
+## Actions
+   - provision
+   - deprovision
+   - update
+   - bind 
+   - unbind
+   
+   
+## Problem Description
+
+Service Bundle's are stateless. All state is managed for them by the broker. Currently the broker passes in parameters specified
+during a request to the catalog. It also passes in some additional parameters such as the namespace etc, additionally 
+we also pass in credentials created during the provision. While this is useful, there is no mechanism for a Service Bundle to store 
+and access data across actions without exposing it to the end user. We want to avoid Service Bundle developers working around this 
+limitation by doing suboptimal things such as storing extra data in the credentials or by creating a ConfigMap in the user's 
+namespace as both of these expose the data to the user and, as the namespace is controlled by the user, it is naturally 
+not trustworthy or reliable.
+
+
+
+# Proposed Solution
+
+## Service Bundle contract for handling state
+
+### Add APB specific module
+
+This module will expose an API to the APB developer for setting key value pairs.
+
+```
+- name: Save some stuff
+  asb_set_state:
+    service_name: "{{ service_name }}"
+
+```
+
+Under the hood this APB module would take the key value pair, and store it in a
+ConfigMap named ```$POD_NAME```. This ConfigMap would live within the 
+temporary namespace ```$POD_NAMESPACE``` and be created by the broker before the APB pod was created. 
+
+## Update broker to manage Service Bundle created state ConfigMaps
+
+To ensure the state is persisted across Service Bundle actions, the broker will create a ConfigMap within the ```$POD_NAMESPACE``` named ```$POD_NAME```. 
+After an action was successfully completed (ie the Service Bundle exited with a 0 exit code) and before the sandbox namespace was removed, the broker would copy the ConfigMap back to the broker's namespace and name it ```<ServiceInstanceID>-state```. If a ConfigMap with 
```

Comment:
If we consider this in a large scale production deployment, are there scaling or quota concerns having potentially 10s of thousands of state maps around?

Reply:
I initially suggested that when the developer calls ```set_state``` it would create a configmap if was not already present. I think it would be better to have lazy creation over eager creation as the requirement for state is not going to be something every apb needs. So happy to go back to that approach. It should significantly reduce the number of objects created although there is still the potential for many objects over a long period of time. 
Is this enough to reduce the concern? I do not know enough about what the upper bound would be for configmaps in a namespace, it is also worth noting that we are not iterating through these objects but instead are accessing them by name. I am assuming that in etcd this is a O(1) action in cost rather than O(n)

Reply:
> I think it would be better to have lazy creation over eager creation as the requirement for state is not going to be something every apb needs.

That's a fair point that will probably dramatically reduce the number of configmaps lying around. :+1: